### PR TITLE
Bento Selector: Imperative API

### PR DESCRIPTION
**Pull request by caroqliu**
Move `toggle` and `clear` to the Preact layer.

Partial for #28282

### CODE REVIEW

File: extensions/amp-selector/1.0/amp-selector.js [hunk 1]
```diff
@@ -50,79 +50,38 @@ class AmpSelector extends PreactBaseElement {
...
     };
 
-    this.registerAction('clear', () => {
-      isExpectedMutation = true;
-      this.mutateProps(dict({'value': []}));
-    });
-
-    this.registerAction('selectUp', (invocation) => {
-      const {args, trust} = invocation;
+    this.registerApiAction('clear', (api, unused) => api./*OK*/ clear());
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/amp-selector.js:53]: The `unused` shouldn't be needed.

File: extensions/amp-selector/1.0/selector.js [hunk 2]
```diff
@@ -75,6 +111,62 @@ export function Selector({
...
+   * the new selected state becomes the option at the given delta.
+   *
+   * ex: (1, [0, 2], [0, 1, 2, 3]) => [2, 1]
+   * ex: (-1, [2, 1], [0, 1, 2, 3]) => [1]
+   * ex: (2, [2, 1], [0, 1, 2, 3]) => [1, 0]
+   * ex: (-1, [], [0, 1, 2, 3]) => [3]
+   */
+  const selectBy = useCallback(
+    (delta) => {
+      const previous = options.indexOf(selected.shift());
```
Comment (dvoytenko) [inline extensions/amp-selector/1.0/selector.js:146]: One thing I'm not quite certain: can we actually rely on any deterministic order with `Selector > Option` structure? Especially in case of mutations? Does it matter to us here? If order matters for this API, we may need to provide `Option({index})` explicitly for this API.

**Review by dvoytenko — CHANGES_REQUESTED** [review]

File: extensions/amp-selector/1.0/selector.js (hunk 2 above)
Comment (caroqliu) [reply extensions/amp-selector/1.0/selector.js:146]: In my observation for the registrations is that they are not cleaned up in order (this is okay, since we remove the first instance of the value) but they _are_ registered in order. Isn't a guarantee of `useEffect` that it is always run in order?

I did manually remove many dom nodes from the options and the actions continue to work as expected. i.e. If we have [1, 2, 3, 4] and we remove [1], [1, 2, 3, 4] all get removed in no clear order, and [2, 3, 4] all get registered in order.

Comment (dvoytenko) [reply extensions/amp-selector/1.0/selector.js:146]: Not sure. Just a bit nervous about this. On the surface - yes, effects all run in order. But what would happen with this component?

```
const [state, setState] = useState(false);
<Selector>
  <Option option="A" />
  {state && (<Option option="B" />)}
  <Option option="C" />
</Selector>
```

We will start with `[A, C]` and when `B` is enabled, we will go to `[A, C, B]`. Is this ok?

File: extensions/amp-selector/1.0/amp-selector.js (hunk 1 above)
Comment (caroqliu) [reply extensions/amp-selector/1.0/amp-selector.js]: Removed

File: extensions/amp-selector/1.0/selector.js (hunk 2 above)
Comment (caroqliu) [reply extensions/amp-selector/1.0/selector.js:146]: You're right, this is not ideal. I don't love the idea of asking for an index when it should be known by its order. What do you think about processing the children instead?

Comment (dvoytenko) [reply extensions/amp-selector/1.0/selector.js:146]: We can't do children because there might be intermediate `<table>`, `<ul>`, `<div>`, etc. I'm actually not sure why we have an index-based APIs at all. But I think if they are needed then either of this should be ok:

1. We indeed process options in order of registration. Somehow user understands that selective showing/hiding of options is going to cause undefined behavior.
2. If that still needed, the user providers explicit `index` or `order` prop.

Basically I'm thinking that the index-based APIs usage is somewhat special and if it's needed, the developer would have to do a bit more to ensure that it works correctly.

Comment (caroqliu) [reply extensions/amp-selector/1.0/selector.js:146]: If the developer has to do more work to ensure index-based APIs works correctly, they could apply a similar amount of work to implement that themselves with the `toggle` api. What do you think about keeping `toggle` and `clear` in Preact layer and implementing `selectUp` and `selectDown` in AMP actions by way of using `toggle`?

We would change the API from `toggle(index, opt_force)` to `toggle(option, opt_force)` so we will not need order in the Preact layer. In the AMP layer we would then get the value from the `option` attribute by DOM order, which we can also manage with the existing MutationObserver. WDYT?

If we'd rather prioritize consistency between the imperative API available between modes, though, then we should go with your approach of (1) status quo with documentation of pitfalls + (2) respecting `index` or `order` if present, which in AMP layer can always be.

Comment (dvoytenko) [reply extensions/amp-selector/1.0/selector.js:146]: Yes. I do like the idea of the index-based APIs being on AMP level only.

Comment (caroqliu) [reply extensions/amp-selector/1.0/selector.js]: Sounds good,  revised the change accordingly, PTAL!

**Review by dvoytenko — APPROVED** [review]
